proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/sdk/impl/GetActivationStatusTask.java

```diff
@@ -0,0 +1,595 @@
+            final Throwable error = new PowerAuthErrorException(PowerAuthErrorCodes.PA2ErrorCodeNetworkError, "Number of upgrade attempts reached its maximum.");
+            completeTask(null, error);
+        }
+    }
+
+    /**
+     * Continue task with the protocol V3 upgrade.
+     *
+     * @param status last received activation status
+     */
+    private void continueWithUpgradeToV3(@NonNull ActivationStatus status) {
+
+        // Keep status for later processing
+        lastFetchedStatus = status;
+
+        // Get server's and ours protocol version.
+        final ProtocolVersion serverVersion = status.currentVersion;
+        final ProtocolVersion localVersion = session.getProtocolVersion();
+
+        if (serverVersion == ProtocolVersion.V2) {
+
+            // Server is still on V2 version, so we need to determine how to continue.
+            // At first, we should check whether the migration was started, because this
+            // continue method must handle all possible migration states.
+
+            if (session.getPendingProtocolUpgradeVersion() == ProtocolVersion.NA) {
+                // Upgrade has not been started yet
+                PA2Log.d("ProtocolUpgrade: Starting activation migration to protocol V3");
+                if (session.startProtocolUpgrade() != ErrorCode.OK) {
+                    completeTaskWithUpgradeError("Protocol upgrade start failed.");
+                    return;
+                }
+                serializeSessionState();
+            }
+
+            // Now lets test the current local protocol version
+            if (localVersion == ProtocolVersion.V2) {
+                // Looks like we didn't start migration on the server, or the request
+                // didn't finish. In other words, we still don't have the CTR_DATA locally.
+                startUpgradeToV3();
+                return;
+
+            } else if (localVersion == ProtocolVersion.V3) {
+                // We already have CTR_DATA, but looks like server didn't receive our "commit" message.
+                // This is because server's version is still in V2.
+                commitUpgradeToV3();
+                return;
+            }
+
+            // Current local version is unknown. This should never happen, unless there's
+            // a new protocol version and migration routine is not updated.
+            // This branch will report "Internal migration error"
+
+        } else if (serverVersion == ProtocolVersion.V3) {
+
+            // Server is already on V3 version, check the local version.
+
+            if (localVersion == ProtocolVersion.V2) {
+                // This makes no sense. Server is in V3, but the client is still in V2.
+                // Only possible explanation is that local session has been restored from some
+                // older backup. For all cases, it's recommended to reset the session.
+                completeTaskWithUpgradeError("Server-Client protocol version mishmash.");
```

Comment:
Typo, should be probably: `mismatch`.

Reply:
Mishmash is a [regular word](https://www.dictionary.com/browse/mishmash) and I think it's more suitable for the situation. In this case, it's a really unexpected mess in the state.